rest-server/engines/src/main/scala/com/actionml/engines/ur/URAlgorithm.scala

```diff
@@ -69,42 +67,25 @@ class URAlgorithm[T] private (initParams: String, dataset: Dataset[T]) extends A
   override def process(): Validated[ValidateError, String] = {
     def myTrainFunction: Iterator[Document] => String = _.mkString(" -- ")
 
-    sparkContext = createSparkContext(
-      appName = dataset.engineId,
-      dbName = dataset.dbName,
-      collection = dataset.collection,
-      config = initParams)
-
-    logger.debug(s"Starting train $this with spark $sparkContext")
-
-    /*
-    sparkContext.andThen { sc =>
-
-      val rdd = sc.makeRDD((1 to 10000).toSeq))
-      val result = rdd.ma
-
-      Valid("URAlgorithm model creation queued for processing on the Spark cluster")
-    }
-    */
-
-    sparkContext.andThen { sc =>
-      val rdd = createRdd(sc)
-      val result = sc.runJob(rdd, myTrainFunction)
-      println("********************************")
-      println(result)
-      println("********************************")
-      Valid(
-        """
-          |{
-          |  "Comment": "Made it to URAlgorithm.train"
-          |  "jobId": "replace with actual Spark + YARN job-id"
-          |  "other": "other useful info"
-          |}
-        """.stripMargin
-      )
-    }
-
-
+    val defaults = Map(
+      "appName" -> dataset.engineId,
+      "deploy-mode" -> "cluster",
+      "spark.mongodb.input.uri" -> MongoStorage.uri,
+      "spark.mongodb.input.database" -> dataset.dbName,
+      "spark.mongodb.input.collection" -> dataset.collection
+    )
+    val result = Await.result(execute(myTrainFunction, initParams, defaults), Duration("5 minutes"))
```

Comment:
How would you use the execute to 
1) return the Valid or Invalid synchronous with the call to `train`
2) run process in the Future thread. 

It looks like the train function must take some iterator, this is not ideal at all, it should be a block of code `{ some code]` that has access to all variable in the scope of the block. There should be no restrictions on input or output. Packaging the input to `execute` as a named function with specific input/output does not seem ideal.

Reply:
we should talk about execute API...